python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+"""
+import traceback
+import copy
+import gc
+import threading
+
+import sgtk
+from sgtk.platform.qt import QtCore
+from sgtk import TankError
+
+class _BackgroundTask(object):
+    """
+    Container class for a single task.
+    """
+    def __init__(self, task_id, cbl, group, priority, args, **kwargs):
+        """
+        Construction.
+
+        :param task_id:     The unique id for this task
+        :param cbl:         Callable to execute to perform the task
+        :param group:       The group that this task belongs to
+        :param priority:    The priority this task should be run with
+        :param args:        Additional arguments that should be passed to func
+        :param kwargs:      Additional named arguments that should be passed to func
+        """
+        self._uid = task_id
+
+        self._cbl = cbl
+        self._args = args
+        self._kwargs = kwargs
+
+        self._group = group
+        self._priority = priority
+
+    def __repr__(self):
+        """
+        Create a string representation of this instance
+        :returns:   A string representation of this instance
+        """
+        return "[%d, G:%s, P:%s] %s" % (self._uid, self._group, self._priority, self._cbl.__name__)
+
+    @property
+    def uid(self):
+        """
+        :returns:   The unique id of this task
+        """
+        return self._uid
+
+    @property
+    def group(self):
+        """
+        :returns:   The group this task belongs to
+        """
+        return self._group
+
+    @property
+    def priority(self):
+        """
+        :returns:   The priority for this task
+        """
+        return self._priority
+
+    def append_upstream_result(self, result):
+        """
+        Append the result from an upstream task to this tasks kwargs.
+
+        :param result:  A dictionary containing the result from an upstream task
+        """
+        if result and isinstance(result, dict):
+            self._kwargs = dict(self._kwargs.items() + result.items())
+
+    def run(self):
+        """
+        Perform this task
+
+        :returns:   The result of performing the task
+        """
+        return self._cbl(*self._args, **self._kwargs)
+
+class _WorkerThreadA(QtCore.QThread):
```

Comment:
It's you ;)

https://github.com/shotgunsoftware/tk-framework-shotgunutils/pull/17/files#diff-8235e4ee893b7034fccfa2c8c7308402R612

I deliberately left _WorkerThreadB in there though in case someone decides to try that recipe later and just hits the same problems I did!